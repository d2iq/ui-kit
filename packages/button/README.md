# Buttons

Buttons are used for action items. Some examples include Create, Save, Delete. Buttons make it obvious to the user that they can take an action.

Do not use buttons as navigational elements. Instead, use anchors because it takes the user to a new page and is not associated with an action.

Each page will likely have one primary button. Any remaining calls-to-action are represented as secondary buttons.

## Button states and types

### Types

**Primary**: Primary call to action button. Should stand out above all other actions. Do not put two primary buttons next to each other. Examples include Save, Deploy.

**Secondary**: Most buttons on a page are likely this style.

**Danger**: Destructive buttons are used to highlight something dangerous that can't be undone e.g. "Delete".

**Icon only**: Use when an icon is enough to represent the action e.g. "Create a Service". Should be accompanied by a tooltip.

**Icon with label**: If there is space and we feel an icon by itself is not obvious enough e.g. "Kill Task"

### States

**Disabled**: Disable a button when the action is not applicable but we want to show the user that it may be possible. Disabled buttons should have a tooltip.

**Processing**: When user clicks or taps a button and an API request is initiated we should provide feedback so they know their action was received.

**Hidden:** Hide a button when there is no way a user can interact with the button.

## Using buttons as links

Sometimes there is a button that is actually a link to somewhere else. If the appropriate props are set, the button is rendered as an anchor tag, but it is styled like a button.

It is not recommended to use a button as a link when it triggers a dropdown. It would be frustrating if a user clicked a button and expected to see a dropdown appear, but they are taken to a new page.

## Button content

Button labels tell users what will happen when they click the button. Use verbs that describe the action. Acceptable labels include:

- New \$object
- Create \$object
- Edit \$object
- Save \$object
- Delete \$object
- Deploy, Scale, Restart

Use sentence-style capitalization (only the first word in a phrase and any proper nouns capitalized) and no more than three words for button labels.

For Sets of Buttons, use specific labels, such as Save or Cancel, instead of using Yes and No. This is particularly helpful when the user is confirming an action.

## ResetButton

A `ResetButton` component is essentially an unstyled HTML `<button>` element. It can be used for an element that is semantically a button, but is not intended to look like a button.

## Feedback

Buttons should have hover, active, focus, disabled states.

When a user interacts with a button, either via click or tap, and an action is being completed, e.g. an API call, the button should change to disabled and provide feedback via the label to the user that it received the interaction and is busy performing the action e.g. “Saving…”

## Tooltips

Buttons may or may not require tooltips to help communicate their intent.

Icon buttons that do not have labels should always have a tooltip so the user knows what will happen when they click it.

Disabled buttons should always have a tooltip to explain why it is disabled so the user is aware.
