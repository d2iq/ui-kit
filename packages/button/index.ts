export { default as PrimaryButton } from "./components/PrimaryButton";
export { default as SecondaryButton } from "./components/SecondaryButton";
export { default as StandardButton } from "./components/StandardButton";
export { default as SuccessButton } from "./components/SuccessButton";
export { default as DangerButton } from "./components/DangerButton";

export { default as PrimaryDropdownButton } from "./components/PrimaryDropdownButton";
export { default as SecondaryDropdownButton } from "./components/SecondaryDropdownButton";
export { default as StandardDropdownButton } from "./components/StandardDropdownButton";
export { default as SuccessDropdownButton } from "./components/SuccessDropdownButton";
export { default as DangerDropdownButton } from "./components/DangerDropdownButton";

export { default as ResetButton } from "./components/ResetButton";
