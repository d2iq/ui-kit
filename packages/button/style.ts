import { css } from "@emotion/css";
import {
  borderRadiusSmall,
  borderRadiusDefault,
  themeBrandPrimary,
  themeSuccess,
  themeError,
  themeTextColorInteractive,
  themeTextColorInteractiveInverted,
  themeTextColorPrimary,
  themeTextColorPrimaryInverted,
  themeBgDisabled,
  themeBgDisabledInverted,
  themeTextColorDisabled,
  themeBgNeutral,
  themeBgNeutralInverted,
  themeBrandPrimaryInverted,
  themeSuccessInverted,
  themeErrorInverted,
  themeTextColorDisabledInverted
} from "../design-tokens/build/js/designTokens";
import { darken, pickReadableTextColor } from "../shared/styles/color";
import { ButtonAppearances } from "./components/ButtonBase";
import { tintContent } from "../shared/styles/styleUtils";
import getCSSVarValue from "../utilities/getCSSVarValue";

export const buttonPadding = {
  vert: "10px",
  horiz: "18px"
};

const getHoverColor = color => darken(color, 1);
const getActiveColor = color => darken(color, 2);

export const filledButton = (
  baseColor: string,
  hoverColor: string,
  activeColor: string,
  isInverse?: boolean
) => {
  const contentColor = isInverse
    ? themeTextColorPrimaryInverted
    : pickReadableTextColor(
        baseColor,
        getCSSVarValue(themeTextColorPrimary),
        getCSSVarValue(themeTextColorPrimaryInverted)
      );

  return css`
    ${tintContent(contentColor)};
    background-color: ${baseColor};
    border-radius: ${borderRadiusSmall};
    padding: ${buttonPadding.vert} ${buttonPadding.horiz};

    &[href],
    &[href]:visited {
      ${tintContent(contentColor)};
    }

    &:hover {
      background-color: ${hoverColor};
    }

    &:active {
      background-color: ${activeColor};
    }
  `;
};

const mutedButton = css`
  ${tintContent(themeTextColorDisabled)};
  cursor: default;
  pointer-events: none;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabled)};
  }
`;

const inverseMutedButton = css`
  ${tintContent(themeTextColorDisabledInverted)};
  cursor: default;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabledInverted)};
  }
`;

const mutedFilledButton = css`
  background-color: ${themeBgDisabled};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabled};
  }
`;

const inverseMutedFilledButton = css`
  background-color: ${themeBgDisabledInverted};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabledInverted};
  }
`;

export const processingTextStyle = css`
  &:after {
    content: "...";
  }
`;

export const buttonBase = css`
  cursor: pointer;
  display: inline-block;
  outline: none;
  text-decoration: none;
`;

export const buttonContent = css`
  outline: none;
`;

export const fullWidthButton = css`
  text-align: center;
  width: 100%;
`;

export const focusStyles = (
  bgColor: string,
  borderColor: string = bgColor
) => css`
  outline: none;
  position: relative;

  &:focus {
    background-color: ${bgColor};

    &:after {
      border: 2px solid ${borderColor};
      border-radius: ${borderRadiusDefault};
      bottom: -3px;
      content: "";
      left: -3px;
      position: absolute;
      right: -3px;
      top: -3px;
    }
  }
`;

export const focusStyleByAppearance = (appearance, isInverse) => {
  switch (appearance) {
    case "primary":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimary)));
    case "secondary":
      return css`
        ${isInverse
          ? focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
            )
          : focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractive))
            )};
        &:focus {
          &:after {
            border-radius: 0;
            border-left-width: 0;
            border-right-width: 0;
            border-top-width: 0;
          }
        }
      `;
    case "standard":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBgNeutralInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBgNeutral)));
    case "success":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeSuccessInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeSuccess)));
    case "danger":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeErrorInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeError)));
    default:
      return "";
  }
};

export const button = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimary),
        getHoverColor(getCSSVarValue(themeBrandPrimary)),
        getActiveColor(getCSSVarValue(themeBrandPrimary))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractive)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractive)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutral),
        getHoverColor(getCSSVarValue(themeBgNeutral)),
        getActiveColor(getCSSVarValue(themeBgNeutral))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccess),
        getHoverColor(getCSSVarValue(themeSuccess)),
        getActiveColor(getCSSVarValue(themeSuccess))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeError),
        getHoverColor(getCSSVarValue(themeError)),
        getActiveColor(getCSSVarValue(themeError))
      );
    default:
      return "";
  }
};

export const buttonInverse = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimaryInverted),
        getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)),
        getActiveColor(getCSSVarValue(themeBrandPrimaryInverted))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractiveInverted)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractiveInverted)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutralInverted),
        getHoverColor(getCSSVarValue(themeBgNeutralInverted)),
        getActiveColor(getCSSVarValue(themeBgNeutralInverted))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccessInverted),
        getHoverColor(getCSSVarValue(themeSuccessInverted)),
        getActiveColor(getCSSVarValue(themeSuccessInverted))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeErrorInverted),
        getHoverColor(getCSSVarValue(themeErrorInverted)),
        getActiveColor(getCSSVarValue(themeErrorInverted))
      );
    default:
      return "";
  }
};

export const getMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${mutedButton};
        ${mutedFilledButton};
      `
    : mutedButton;
};

export const getInverseMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${inverseMutedButton};
        ${inverseMutedFilledButton};
      `
    : inverseMutedButton;
};

export const pointerCursor = css`
  cursor: pointer;
`;

// Replicates default browser focus ring styles.
//
// The media query targets Webkit browsers, which can
// more accurately replicate the native focus ring style
export const keyboardFocus = css`
  > div:focus {
    outline: none;
  }

  &:focus > div {
    outline-color: Highlight;
    outline-width: thin;

    @media (-webkit-min-device-pixel-ratio: 0) {
      outline-color: -webkit-focus-ring-color;
      outline-style: auto;
      outline-width: unset;
    }
  }
`;
