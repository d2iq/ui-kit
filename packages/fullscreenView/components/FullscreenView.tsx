import * as React from "react";
import { cx, css } from "@emotion/css";
import { ButtonProps } from "../../button/components/ButtonBase";
import { flex, padding, flexItem } from "../../shared/styles/styleUtils";
import { modalContent, fullscreenModalHeader } from "../style";
import FullscreenViewHeader from "./FullscreenViewHeader";

export interface FullscreenViewProps {
  /** The primary button */
  ctaButton?: React.ReactElement<ButtonProps>;
  /** The text for the secondary button that closes the view */
  closeText: React.ReactNode;
  /** The title that appears in the header */
  title: React.ReactNode;
  /** The subtitle that appears in the header */
  subtitle?: React.ReactNode;
  /** Whether we automatically add padding to the body of the modal. */
  isContentFlush?: boolean;
  /** Custom header content component. ⚠️Use rarely and with caution⚠️ */
  headerComponent?: React.JSXElementConstructor<any>;
  /** Custom banner component that will appear above the header. */
  bannerComponent?: React.ReactNode;
  /** Function that gets called when the modal is closed */
  onClose: (event?: React.SyntheticEvent<HTMLElement>) => void;
  /** The base of the `data-cy` value. This is used for writing selectors in Cypress. */
  cypressSelectorBase?: string;
  /** The css max width around the header contents  */
  containerMaxWidth?: string;
  children?: React.ReactNode;
}

const FullscreenView = ({
  children,
  ctaButton,
  closeText,
  isContentFlush,
  onClose,
  title,
  subtitle,
  bannerComponent,
  headerComponent,
  cypressSelectorBase = "fullscreenView",
  containerMaxWidth
}: FullscreenViewProps) => {
  const HeaderComponent = headerComponent ?? FullscreenViewHeader;

  let container: string = "";
  if (containerMaxWidth) {
    // add border-box to account for any left and right padding
    container = css`
      margin-left: auto;
      margin-right: auto;
      max-width: ${containerMaxWidth};
      box-sizing: border-box;
    `;
  }

  return (
    <div className={flex({ direction: "column" })}>
      <div
        className={cx(fullscreenModalHeader)}
        data-cy={`${cypressSelectorBase}-header`}
      >
        {bannerComponent && (
          <div data-cy={`${cypressSelectorBase}-banner`}>{bannerComponent}</div>
        )}
        <div className={cx(padding("all", "xl"), container)}>
          <HeaderComponent
            title={title}
            subtitle={subtitle}
            ctaButton={ctaButton}
            closeText={closeText}
            onClose={onClose}
          />
        </div>
      </div>
      <div
        className={cx(modalContent, flexItem("grow"), {
          [padding("all", "xl")]: !isContentFlush
        })}
        data-cy={`${cypressSelectorBase}-content`}
      >
        {children}
      </div>
    </div>
  );
};

export default React.memo(FullscreenView);
