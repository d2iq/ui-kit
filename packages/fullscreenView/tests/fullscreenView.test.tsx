import React from "react";
import { mount } from "enzyme";
import serializer from "@emotion/jest";
import toJson from "enzyme-to-json";
import FullscreenView from "../components/FullscrenView";
import { PrimaryButton } from "../../button";

expect.addSnapshotSerializer(serializer);

describe("FullscreenView", () => {
  it("renders FullscreenView", () => {
    const closeAction = jest.fn();
    const component = mount(
      <FullscreenView
        onClose={closeAction}
        ctaButton={<PrimaryButton>CTA</PrimaryButton>}
        closeText="Close"
        title="Title"
      >
        <div tabIndex={0}>I am modal content</div>
      </FullscreenView>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
