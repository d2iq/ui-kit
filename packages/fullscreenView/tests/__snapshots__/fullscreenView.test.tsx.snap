// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullscreenView renders FullscreenView 1`] = `
.emotion-0 {
  height: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-0>div {
  width: 100%;
}

.emotion-1 {
  background-color: var(--themeBgSecondary, #F7F8F9);
  box-sizing: border-box;
  border-bottom: 1px solid #DADDE2;
  padding: 32px;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
}

.emotion-2>div {
  width: auto;
}

.emotion-3 {
  text-align: left;
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
}

.emotion-4 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: var(--themeTextColorInteractive, #7D58FF);
  fill: var(--themeTextColorInteractive, #7D58FF);
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
}

.emotion-4::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-4:hover {
  color: #704fe5;
  fill: #704fe5;
}

.emotion-4:active {
  color: #6446cc;
  fill: #6446cc;
}

.emotion-4[href],
.emotion-4[href]:visited {
  color: var(--themeTextColorInteractive, #7D58FF);
  fill: var(--themeTextColorInteractive, #7D58FF);
}

.emotion-5 {
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
}

.emotion-6 {
  text-align: center;
  margin-top: 0;
  margin-bottom: 0;
  padding-bottom: 8px;
  font-size: 18px;
  font-weight: 400;
}

.emotion-7 {
  text-align: center;
  margin-top: 0;
  margin-bottom: 0;
  font-size: 12px;
  font-weight: 400;
  color: #76797E;
  fill: #76797E;
}

.emotion-8 {
  text-align: right;
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
}

.emotion-9 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #7D58FF;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
}

.emotion-9::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-9[href],
.emotion-9[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-9:hover {
  background-color: #704fe5;
}

.emotion-9:active {
  background-color: #6446cc;
}

.emotion-10 {
  box-sizing: border-box;
  overflow: auto;
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
  padding: 32px;
}

<FullscreenView
  closeText="Close"
  ctaButton={
    <PrimaryButton>
      CTA
    </PrimaryButton>
  }
  onClose={[MockFunction]}
  title="Title"
>
  <div
    className="emotion-0"
  >
    <div
      className="emotion-1"
      data-cy="fullscreenView-header"
    >
      <Delegate
        default={[Function]}
        passDefault={true}
        props={
          Object {
            "closeText": "Close",
            "ctaButton": <PrimaryButton>
              CTA
            </PrimaryButton>,
            "onClose": [MockFunction],
            "subtitle": undefined,
            "title": "Title",
          }
        }
      >
        <FullscreenModalHeader
          closeText="Close"
          ctaButton={
            <PrimaryButton>
              CTA
            </PrimaryButton>
          }
          onClose={[MockFunction]}
          title="Title"
        >
          <div
            className="emotion-2"
          >
            <div
              className="emotion-3"
            >
              <SecondaryButton
                onClick={[MockFunction]}
              >
                <ButtonBase
                  appearance="secondary"
                  data-cy="secondaryButton"
                  onClick={[MockFunction]}
                >
                  <FocusStyleManager
                    focusEnabledClass="css-1h85cd2"
                  >
                    <button
                      className="emotion-4"
                      data-cy="secondaryButton"
                      onClick={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span>
                        Close
                      </span>
                    </button>
                  </FocusStyleManager>
                </ButtonBase>
              </SecondaryButton>
            </div>
            <div
              className="emotion-5"
            >
              <h2
                className="emotion-6"
              >
                Title
              </h2>
              <h3
                className="emotion-7"
              />
            </div>
            <div
              className="emotion-8"
            >
              <PrimaryButton>
                <ButtonBase
                  appearance="primary"
                  data-cy="primaryButton"
                >
                  <FocusStyleManager
                    focusEnabledClass="css-xcotv6"
                  >
                    <button
                      className="emotion-9"
                      data-cy="primaryButton"
                      onClick={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span>
                        CTA
                      </span>
                    </button>
                  </FocusStyleManager>
                </ButtonBase>
              </PrimaryButton>
            </div>
          </div>
        </FullscreenModalHeader>
      </Delegate>
    </div>
    <div
      className="emotion-10"
      data-cy="fullscreenView-content"
    >
      <div
        tabIndex={0}
      >
        I am modal content
      </div>
    </div>
  </div>
</FullscreenView>
`;
